### ISSUE TITLE
VSCode extension

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#11087]
> Many developers are now on vs code and it's ruling IDE market according to stackoverflow survey 2019. I wish there would be conan extension for vs code.
> 
> https://insights.stackoverflow.com/survey/2019#technology-_-most-popular-development-environments

@KaungZawHtet, @jsgogo  Maybe it's a bit late to reply this thread, but I created the conan extension for VS code, and just published it a couple days ago. Maybe you can check out my extension :) I hope it could be really helpful for the community :)
Here is the link to my extension https://github.com/afri-bit/vsconan

_Originally posted by @afri-bit in https://github.com/conan-io/wishlist/issues/226#issuecomment-1107853636_

**Comment by memsharded** [comment]
Hi @afri-bit

I am creating a new issue here in the main Conan repo, because the ``wishlist`` is mainly for packages, and is almost abandoned in favor of the main ``conan-center-index`` repo.

Indeed an extension for VScode is interesting, and we have been wanting to do something for it for a while. At the moment we are prioritizing getting 2.0 out, but after 2.0 we will resume work on IDE plugins and extensions. We will continue the conversation here when it is time, so thanks very much for doing this and sharing it!

**Comment by afri-bit** [comment]
Hi @memsharded,

thanks for creating a new issue for this topic.

Looking forward to Conan 2.0 :)
I hope the CLI API will not change so much at the end, so my extension can be reused, since it just uses CLI mainly for this purpose.

Other than that, always happy to help :)

**Comment by memsharded** [comment]
> I hope the CLI API will not change so much at the end, so my extension can be reused, since it just uses CLI mainly for this purpose.

Unfortunately it will change substantially, the Conan 1.X cli accumulated some bad UX, so it was time for a clean up. But ``conan install <path> -pr=myprofile`` remains, it depends on how many things you use. But don't worry now, we will figure it out when it comes.

**Comment by tbsuht** [comment]
> Indeed an extension for VScode is interesting, and we have been wanting to do something for it for a while.

*Push*. This would be great!

See also: https://github.com/conan-io/conan/issues/7374

**Comment by jcar87** [comment]
> > Indeed an extension for VScode is interesting, and we have been wanting to do something for it for a while.
> 
> _Push_. This would be great!
> 
> See also: #7374

Hi @tbsuht  - in the meantime, if your project uses CMake and already has a `conanfile` (txt or py), you may want to look into https://github.com/conan-io/cmake-conan - Using VSCode's CMake tools extension, all you would need to do is edit the presets to define `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` to point to `[path-to-cmake-conan]/conan_provider.cmake` - this can be persisted in `CMakePresets.json` or `CMakeUserPresets.json` at the root of your project. This should cause CMake to invoke Conan while it configures - negating the need to call Conan beforehand (although it has some limitations, please see that project's readme).

**Comment by tbsuht** [comment]
Hi @jcar87 

We're already using `CMake-Conan` and hitting the limitations there. In our case it's information from the cpp_info (`builddirs`) of an internal package made available via `CMakeToolchain` generator which can't be used when building.

Is there any plan for a plugin? Can't find something in the milestone section.

**Comment by memsharded** [comment]
I think the experience with VSCode with the ``cmake-tools`` installed using presets is quite good. Yes, it requires doing ``conan install`` in the terminal, but the presets are automatically loaded and can be used from VSCode, so the effort is smaller than other IDEs. 

At the moment a VSCode plugin is not planned.

**Comment by tbsuht** [comment]
@memsharded Sure, the integration into VS Code is good. The problem is that the `CMakeToolchain` generator is not supported, to my knowledge. And this limitation is easily reachable imo, see the example above.

**Comment by memsharded** [comment]
> Sure, the integration into VS Code is good. The problem is that the CMakeToolchain generator is not supported, to my knowledge. And this limitation is easily reachable imo, see the example above.

Yes, the ``CMakeToolchain`` is supported, and it is the one that generates the presets and toolchain for the integration to work. I am referring to the flow using ``conan install .`` + VSCode cmake-tools integration opening the presets generated by Conan, this works quite well, with the ``CMakeToolchain``. The only thing is that user should type ``conan install .``.

**Comment by tbsuht** [comment]
Ok, I wasn't precise enough. :)

I was talking about the one-click-to-build solution without using `conan install`.

**Comment by jcar87** [comment]
> Hi @jcar87
> 
> We're already using `CMake-Conan` and hitting the limitations there. In our case it's information from the cpp_info (`builddirs`) of an internal package made available via `CMakeToolchain` generator which can't be used when building.
> 
> Is there any plan for a plugin? Can't find something in the milestone section.

Hi @tbsuht - if you could open an issue with more details in the `cmake-conan` repository we may be able to advise further, if anything so that we can have a better view of what is missing feature-wise :)

**Comment by listepo** [comment]
Any news?

**Comment by memsharded** [comment]
Hi @listepo 

So far this doesn't seem worth it.
The basic case above:

> Yes, the CMakeToolchain is supported, and it is the one that generates the presets and toolchain for the integration to work. I am referring to the flow using conan install . + VSCode cmake-tools integration opening the presets generated by Conan, this works quite well, with the CMakeToolchain. The only thing is that user should type conan install ..

Works well, and it is just one ``conan install`` command, so flow is mostly ``git clone + conan install + vscode (cmake-tools)``, relatively straightforward.

For other cases, the ``cmake-conan`` integration can automatically call ``conan install`` when cmake is called, and that satisfies other IDE integrations, for example it is used in CLion too.

It would be great to get a more detailed description of the specific pain points, the developer flows that could be improved, what are the expectations from such an extension, etc.

**Comment by ericLemanissier** [comment]
Here is my use case for a conan extension:
I have a recipe with several build requirements tools (cmake, ninja, [the compiler](https://developer.arm.com/tools-and-software/gnu-toolchain), and [a build framework](https://open-cmsis-pack.github.io/cmsis-toolbox/)). Building from console using `conan create` works as expected. 
The pain comes when using vscode: the build requirements are simply not available. There is currently no way to publish the virtual build environment containing the build requirements, so that they can be used by other vscode extensions (like https://marketplace.visualstudio.com/items?itemName=Arm.cmsis-csolution). I can declare a custom task calling `conan install -c tools.env:dotenv=True` unfortunately there is no way to publish the environment variables to the whole vscode environment, or at least to all terminals started after this.

The current alternative to this is unfortunately to use vcpkg and [this extension](https://marketplace.visualstudio.com/items?itemName=Arm.environment-manager): it is able to download vcpkg artifacts for these 4 build requirements, and inject them in the environment. The typescript source code of this extension seems to be closed, but the javascript code shows that modifying `environmentVariableCollection` publishes environment variables

In the best of the world, a conan vscode extension would notice the conanfile.py/.txt, generate the virtual build environment (possibly taking command line argument from user/workslpace settings), and then modify vscode environment with the environment generated by conan.

<details><summary>AI take on this</summary>
<p>
custom vscode task running

`conan install -c tools.env:dotenv=True && cat build\\conan\\conanbuildenv*.env > .env`

combined with this vscode extension

``` javascript
const vscode = require('vscode');
const fs = require('fs');
const path = require('path');

function activate(context) {
    const envCollection = context.environmentVariableCollection;

    function loadEnvFile() {
        const workspaceFolders = vscode.workspace.workspaceFolders;
        if (!workspaceFolders) {
            return;
        }

        const workspacePath = workspaceFolders[0].uri.fsPath;
        const envFilePath = path.join(workspacePath, '.env');

        if (!fs.existsSync(envFilePath)) {
            return;
        }

        const envContent = fs.readFileSync(envFilePath, 'utf8');
        const lines = envContent.split('\n');

        for (const line of lines) {
            const trimmedLine = line.trim();
            
            if (!trimmedLine || trimmedLine.startsWith('#')) {
                continue;
            }

            const match = trimmedLine.match(/^([^=]+)=(.*)$/);
            if (match) {
                const key = match[1].trim();
                let value = match[2].trim();

                if ((value.startsWith('"') && value.endsWith('"')) || 
                    (value.startsWith("'") && value.endsWith("'"))) {
                    value = value.slice(1, -1);
                }
                envCollection.replace(key, value);
                process.env[key] = value; // Also set it in the process environment
            }
        }
    }
    loadEnvFile();

    const envWatcher = vscode.workspace.createFileSystemWatcher('**/.env');
    envWatcher.onDidChange(loadEnvFile);
    envWatcher.onDidCreate(loadEnvFile);
    envWatcher.onDidDelete(() => {
        envCollection.clear();
    });

    const reloadCommand = vscode.commands.registerCommand('envFileLoader.reload', loadEnvFile);

    context.subscriptions.push(envWatcher, reloadCommand);
}

function deactivate() {
}

module.exports = {
    activate,
    deactivate
};
```
</p>
</details>

**Comment by memsharded** [comment]
Thanks very much for the feedback @ericLemanissier 

I think that makes a lot of sense. It is not a "Conan extension" per-se, in the sense that the main goal is not necessarily to call ``conan install`` commands. Instead, the main goal is the activation of the environment that will make the tools installed by Conan available generally to the full VSCode.

I think this is a direction that it is worth investigating.

**Comment by ericLemanissier** [comment]
It looks like the snippet I pasted in my last message works only after executing task `Developer: Reload Window`.
After reading https://github.com/microsoft/vscode/issues/152806 it looks like there is currently no way for an extension to push environment variables to all other extensions, the only work around seems to be that some extension pull environment from other specific extensions. 
I'm not sure how https://github.com/Open-CMSIS-Pack/vscode-cmsis-solution pulls from https://marketplace.visualstudio.com/items?itemName=Arm.environment-manager, but references it at [multiple places](https://github.com/search?q=repo%3AOpen-CMSIS-Pack%2Fvscode-cmsis-solution+environment-manager+language%3ATypeScript&type=code&l=TypeScript)
